# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/android/TraversalRule.h
# Commit: c38e72058869
# Full Hash: c38e720588697e5c0917cab88a4fea6d70af71e1
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2019-09-24 03:42:17
# Regressor Bug: 1564549
# File Overlap Count: 1
# Description:
#   Bug 1564549 - Introduce native traveral rule. r=yzen
#   
#   The rule takes a granularity constant as an argument in the constructor.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D45598
# ==============================================================================

diff -r aca91e9add16 -r c38e72058869 accessible/android/TraversalRule.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/accessible/android/TraversalRule.h	Mon Sep 23 21:06:49 2019 +0000
@@ -0,0 +1,54 @@
+/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef _TraversalRule_H_
+#define _TraversalRule_H_
+
+#include "Pivot.h"
+
+namespace mozilla {
+namespace a11y {
+
+class Accessible;
+
+/**
+ * Class represents a simple traversal rule.
+ */
+class TraversalRule final : public PivotRule {
+ public:
+  TraversalRule();
+  explicit TraversalRule(int32_t aGranularity);
+
+  ~TraversalRule() = default;
+
+  virtual uint16_t Match(Accessible* aAccessible) override;
+
+ private:
+  bool IsSingleLineage(Accessible* aAccessible);
+
+  bool IsFlatSubtree(const Accessible* aAccessible);
+
+  bool IsListItemBullet(const Accessible* aAccessible);
+
+  bool HasName(const Accessible* aAccessible);
+
+  uint16_t DefaultMatch(Accessible* aAccessible);
+
+  uint16_t LinkMatch(Accessible* aAccessible);
+
+  uint16_t HeadingMatch(Accessible* aAccessible);
+
+  uint16_t ControlMatch(Accessible* aAccessible);
+
+  uint16_t SectionMatch(Accessible* aAccessible);
+
+  int32_t mGranularity;
+};
+
+}  // namespace a11y
+}  // namespace mozilla
+
+#endif
\ No newline at end of file